pkg/attributes/attributes.go

```diff
@@ -0,0 +1,368 @@
+)
+
+// Attributes provides a way to add a map of arbitrary YAML/JSON
+// objects.
+// +kubebuilder:validation:Type=object
+// +kubebuilder:validation:XPreserveUnknownFields
+type Attributes map[string]apiext.JSON
+
+// MarshalJSON implements custom JSON marshaling
+// to support free-form attributes
+func (attributes Attributes) MarshalJSON() ([]byte, error) {
+	return json.Marshal(map[string]apiext.JSON(attributes))
+}
+
+// UnmarshalJSON implements custom JSON unmarshaling
+// to support free-form attributes
+func (attributes *Attributes) UnmarshalJSON(data []byte) error {
+	return json.Unmarshal(data, (*map[string]apiext.JSON)(attributes))
+}
+
+// Exists returns `true` if the attribute with the given key
+// exists in the attributes map.
+func (attributes Attributes) Exists(key string) bool {
+	_, exists := attributes[key]
+	return exists
+}
+
+// GetString allows returning the attribute with the given key
+// as a string. If the attribute JSON/YAML content is
+// not a JSON string, then the result will be the empty string
+// and an error is raised.
+//
+// An optional error holder can be passed as an argument
+// to receive any error that might have be raised during the attribute
+// decoding
+func (attributes Attributes) GetString(key string, errorHolder ...*error) string {
+	if attribute, exists := attributes[key]; exists {
+		result := new(string)
+		err := json.Unmarshal(attribute.Raw, result)
+		if err == nil {
+			return *result
+		}
+		if len(errorHolder) > 0 && errorHolder != nil {
+			*errorHolder[0] = err
+		}
+	}
+	return ""
+}
+
+// GetNumber allows returning the attribute with the given key
+// as a float64. If the attribute JSON/YAML content is
+// not a JSON number, then the result will be the zero value
+// and an error is raised.
+//
+// An optional error holder can be passed as an argument
+// to receive any error that might have be raised during the attribute
+// decoding
+func (attributes Attributes) GetNumber(key string, errorHolder ...*error) float64 {
+	if attribute, exists := attributes[key]; exists {
+		result := new(float64)
+		err := json.Unmarshal(attribute.Raw, result)
+		if err == nil {
+			return *result
+		}
+		if len(errorHolder) > 0 && errorHolder != nil {
+			*errorHolder[0] = err
+		}
+	}
+	return 0
+}
+
+// GetBoolean allows returning the attribute with the given key
+// as a bool. If the attribute JSON/YAML content is
+// not a JSON boolean, then the result will be the `false` zero value
+// and an error is raised.
+//
+// An optional error holder can be passed as an argument
+// to receive any error that might have be raised during the attribute
+// decoding
+func (attributes Attributes) GetBoolean(key string, errorHolder ...*error) bool {
+	if attribute, exists := attributes[key]; exists {
+		result := new(bool)
+		err := json.Unmarshal(attribute.Raw, result)
+		if err == nil {
+			return *result
+		}
+		if len(errorHolder) > 0 && errorHolder != nil {
```

Comment:
IDE suggests that `Condition is always 'true' because 'errorHolder' is always not 'nil'` here and in other places in this file.

Reply:
Good catch ! Thanks !

Reply:
Fixed in commit https://github.com/devfile/api/pull/214/commits/ed4b1865c5d7ee354d764eb777f116dba96af9d4